### ISSUE TITLE
Qwen ring matmul failing PCC check

### ISSUE THREAD

**Issue by johanna-rock-tt** [tenstorrent/tt-metal#34413]
Qwen model produces bad PCC / bad accuracy. The issue can be reproduced by a 3 layer model run comparing to the reference output after each operation -> this shows that the FF1 produces bad PCC.

Repro instructions for the whole 3 layer model run:
- branch: `jrock/qwen-debug-ref-comp`
- machine: galaxy
- command: `HF_MODEL=Qwen/Qwen3-32B TT_CACHE_PATH=/proj_sw/user_dev/Qwen3-32B/model_cache/Qwen/Qwen3-32B pytest models/demos/llama3_70b_galaxy/tests/module_tests/test_qwen_transformer.py::test_qwen_transformer_inference`

Loading the input and weight from that FF1 ring matmul into a separate test and running a normal ttnn.matmul works fine.

Repro normal matmul with loaded inputs/weights:
- command: `pytest models/demos/llama3_70b_galaxy/tests/test_matmul.py`

**Comment by johanna-rock-tt** [comment]
Next steps on models side:
- run ring matmul with loaded inputs
- run prefetcher+ ring matmul with loaded inputs

**Comment by alingTT** [comment]
After fixing both the test_ring_matmul  and test_prefetcher_ring_matmul unit test - the ring matmul with or without prefetcher has PCC 1.0 with the normal matmul so this indicates that it is indeed something wrong with the matmul, not the prefetcher interaction

Unit test fixes has been pushed to branch `jrock/qwen-debug-ref-comp`, some PCC values to show

```
test_matmul:
Non-Prefetch Matmul PCC with reference: 0.9839525228158958 <- bad PCC
(False, 'Max ATOL Delta: 9.207382202148438, Max RTOL Delta: inf')
Non-Ring Matmul PCC with Ring matmul output: 0.9996753569862868 <- ring matmul output from the model
(False, 'Max ATOL Delta: 1.375, Max RTOL Delta: nan')
test_ring_matmul:
Non-Prefetch Ring Matmul PCC with torch reference: 0.98449138984136 <- bad PCC
(False, 'Max ATOL Delta: 9.082382202148438, Max RTOL Delta: inf')
Non-Prefetch Ring Matmul PCC with Prefetcher Ring Matmul output: 1.0 <- also ring matmul output from the model
(True, 'Max ATOL Delta: 0.0, Max RTOL Delta: nan')
test_prefetcher_ring_matmul:
Prefetcher Ring Matmul PCC with reference: 0.98449138984136 <- still bad PCC
(False, 'Max ATOL Delta: 9.082382202148438, Max RTOL Delta: inf')
Prefetcher Ring Matmul PCC with Prefetcher Ring Matmul output from model: 1.0
(True, 'Max ATOL Delta: 0.0, Max RTOL Delta: nan')
```

Next steps:
- investigate inside the kernel why are seeing this drop in PCC, putting DPRINTs of the intermediates and review ring all gather logic, grid sizes core selection logic etc

**Comment by alingTT** [comment]
We have a minimal unit test `test_matmul_1d_minimal.py` pushed to branch `aling/ring-mm-bad`
This unit test loads the multi device tensor of in0 and in1 and runs every chunk with Ring MM and Plain MM and does a PCC comparison with torch. This uses the same program configs and memory configs in the model for FF1 in MLP.  Ring MM seems to be faulty as it fails PCC while plain MM gives good PCC.

To reproduce:
`git checkout aling/ring-mm-bad`
`git lfs pull` <- to pull the weight files
`git lfs fetch --include="models/demos/llama3_70b_galaxy/tests/*.pt"` <- to specifically fetch the pt files

Run: `pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_1d_minimal.py`

Note:
- if you comment out running plain matmul and just run ring matmul with torch you will see bad PCC on only the odd column device chunks, if you run both plain matmul and ring matmul you will see bad PCC for the ring MM vs torch on all device chunks
- this unit test assumes an unharvested system with a 7x10 compute grid

**Comment by johanna-rock-tt** [comment]
@bbradelTT is looking at the unit test

**Comment by johanna-rock-tt** [comment]
In parallel we are working on replacing the ring matmuls with normal matmuls (no prefetcher) to validate that except for the ring_matmul issue we are then seeing good accuracy/eval scores.

**Comment by bbradelTT** [comment]
@alingTT said that with program caching disabled pcc is good everywhere.

I move the grid coordinates with y=9 to y=8 and that worked on an x1.

Same behaviour.

```
device.disable_and_clear_program_cache()
```
made things good.

Without that, first pcc was good, rest were bad.

Seems like the input tensor addresses are not updated.

However, @alingTT said that even with program cache off, the model still has bad pcc.

**Comment by alingTT** [comment]
Removing the prefetcher in all of the ring matmuls gives good PCC in the `test_qwen_transformer` test, the branch has been pushed to `aling/qwen-no-pf` , this branch has program caching disabled as well since it is known that ring matmul is problematic with program caching

Next steps:
- investigate a minimal unit test to reproduce this bad PCC with prefetcher and calling the prefetcher op multiple times with qwen mlp shapes 
- run evals on branch provided to make sure this fixes our bad accuracy tests

**Comment by bbradelTT** [comment]
I simplified the test to not need to read data and to run even on an N300 by moving row 9 to row 3. The test is:
```
import pytest
import torch
import ttnn
from loguru import logger
from models.common.utility_functions import is_blackhole
from tests.tt_eager.python_api_testing.sweep_tests.comparison_funcs import comp_pcc
from tests.ttnn.utils_for_testing import assert_with_pcc
from tracy import signpost

# Hardcoded grid from PREFETCHER_NOC1_OUTPUT_GRID
GRID = [
    (1, 3),
    (2, 3),
    (1, 0),
    (2, 0),
    (1, 4),
    (2, 4),
    (1, 5),
    (2, 5),
    (5, 0),
    (6, 0),
    (5, 3),
    (6, 3),
    (5, 1),
    (6, 1),
    (5, 7),
    (6, 7),
    (5, 6),
    (6, 6),
    (5, 2),
    (6, 2),
    (5, 4),
    (6, 4),
    (5, 5),
    (6, 5),
]


def run_matmul_1d_dram_sharded(device, num_iters=1):
    """
    Minimal matmul 1D test with all configs hardcoded from debugger output.
    Compares ring matmul vs plain matmul vs torch output.
    """
    # Core range set for in0 and output
    #device.disable_and_clear_program_cache()
    core_range_set = ttnn.CoreRangeSet([ttnn.CoreRange(ttnn.CoreCoord(x, y), ttnn.CoreCoord(x, y)) for x, y in GRID])

    # in0 sharded in L1: shape={32, 64}
    in0_sharded_mem_config = ttnn.MemoryConfig(
        ttnn.TensorMemoryLayout.WIDTH_SHARDED,
        ttnn.BufferType.L1,
        ttnn.ShardSpec(core_range_set, [32, 64], ttnn.ShardOrientation.ROW_MAJOR),
    )

    # in1 sharded in DRAM: grid={(x=0,y=0)-(x=11,y=0)}, shape={1280, 320}
    in1_shard_grid = ttnn.CoreRangeSet({ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(11, 0))})
    in1_sharded_mem_config = ttnn.MemoryConfig(
        ttnn.TensorMemoryLayout.WIDTH_SHARDED,
        ttnn.BufferType.DRAM,
        ttnn.ShardSpec(in1_shard_grid, [1280, 320], ttnn.ShardOrientation.ROW_MAJOR),
    )

    # Output sharded in L1 (same grid as in0)
    output_sharded_mem_config = ttnn.MemoryConfig(
        ttnn.TensorMemoryLayout.WIDTH_SHARDED,
        ttnn.BufferType.L1,
        ttnn.ShardSpec(core_range_set, [32, 160], ttnn.ShardOrientation.ROW_MAJOR),
    )

    # Program config for ring matmul
    program_config = ttnn.MatmulMultiCoreReuseMultiCast1DProgramConfig(
        compute_with_storage_grid_size=(6, 4),
        in0_block_w=1,
        out_subblock_h=1,
        out_subblock_w=5,
        per_core_M=1,
        per_core_N=5,
        fuse_batch=True,
        fused_activation=None,
        mcast_in0=False,
        gather_in0=True,
        hop_cores=ttnn.CoreRangeSet([]),
        untilize_out=False,
    )

    compute_kernel_config = ttnn.init_device_compute_kernel_config(
        device.arch(),
        math_fidelity=ttnn.MathFidelity.HiFi2,
        math_approx_mode=True,
        fp32_dest_acc_en=True,
        packer_l1_acc=True,
        dst_full_sync_en=True,
    )

    # Create test data
    torch.manual_seed(0)
    in0_original = torch.randn([1,1,32,1280], dtype=torch.bfloat16)
    in1_original = torch.randn([1,1,1280,3200], dtype=torch.bfloat16)
    logger.info(f"IN0 {in0_original}\nIN1 {in1_original}")

    for _ in range(3):
            in0 = in0_original[0, 0].unsqueeze(0).unsqueeze(0)
            in1 = in1_original[0, 0].unsqueeze(0).unsqueeze(0)

            # --- Ring matmul ---
            in0_t = ttnn.from_torch(
                in0,
                device=device,
                layout=ttnn.TILE_LAYOUT,
                dtype=ttnn.bfloat16,
                memory_config=in0_sharded_mem_config,
            )
            in1_t = ttnn.from_torch(
                in1,
                device=device,
                layout=ttnn.TILE_LAYOUT,
                dtype=ttnn.bfloat8_b,
                memory_config=in1_sharded_mem_config,
            )

            signpost("start ring")
            for _ in range(num_iters):
                ring_output_t = ttnn.matmul(
                    in0_t,
                    in1_t,
                    program_config=program_config,
                    memory_config=output_sharded_mem_config,
                    compute_kernel_config=compute_kernel_config,
                    dtype=ttnn.bfloat16,
                )
            logger.info(f"RING_OUT {ring_output_t}")
            signpost("stop ring")
            ring_out = ttnn.to_torch(ring_output_t)

            # --- Plain matmul (like test_matmul.py) ---
            in0_plain = ttnn.from_torch(
                in0,
                device=device,
                layout=ttnn.TILE_LAYOUT,
                dtype=ttnn.bfloat16,
                memory_config=ttnn.L1_MEMORY_CONFIG,
            )
            in1_plain = ttnn.from_torch(
                in1,
                device=device,
                layout=ttnn.TILE_LAYOUT,
                dtype=ttnn.bfloat8_b,
                memory_config=ttnn.DRAM_MEMORY_CONFIG,
            )

            logger.info(f"IN0P {in0_plain}\nIN1P {in1_plain}")

            # --- Torch reference ---
            torch_out = in0 @ in1.to(torch.bfloat16)

            # --- PCC comparisons ---
            passing_ring_vs_torch, pcc_ring_vs_torch = comp_pcc(ring_out, torch_out, 0.98)

            logger.info(f"  Ring vs Torch:        {pcc_ring_vs_torch}")
            assert_with_pcc(ring_out, torch_out, 0.9)


@pytest.mark.skipif(is_blackhole(), reason="Test suite for WH only")
@pytest.mark.parametrize(
    "device_params",
    [{"dispatch_core_axis": ttnn.DispatchCoreAxis.COL}],
    indirect=True,
)
def test_multi_core_matmul_1d_wh_minimal(device, function_level_defaults):
    run_matmul_1d_dram_sharded(device)
```

I found that if I don't create the plain tensors, the test passes.

**Comment by alingTT** [comment]
Update: 
Increasing the global cb size to 800 tiles allows the `test_qwen_transformer` test to pass, so we do not have to disable prefetcher
Will continue to investigate root cause for why 728 tiles is not sufficient and why that causes memory to be overwritten or global cb does not present correct values

Next steps:
- publish new branch and rerun evals

**Comment by bbradelTT** [comment]
For the program cache issue, the problem is that
ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_1d_program_factory.cpp ~l.2265
has
```
        std::vector<uint32_t> mm_in1_args = {
            (std::uint32_t)core_type,
            in1_buffer->address(),  // in1_tensor_addr
            i,                      // ring_idx
        };
```
to set the runtime arg while override_gather_in0_program_parameters has
```
    if (not src1_sharded) {
        auto& writer_runtime_args_by_core = GetRuntimeArgs(program, override_variables.kernels.at(0));
        for (uint32_t i = 0; i < override_variables.cores.size(); ++i) {
            const auto& core = override_variables.cores[i];
            auto& writer_runtime_args = writer_runtime_args_by_core[core.x][core.y];

            /* in1 */
            writer_runtime_args[1] = src_buffer_b->address();
        }
    }
```
That if statement condition will is incorrect.

I will explore using
```
bool in1_uses_buffer = !src1_sharded || (src1_sharded && src_buffer_b->is_dram());
if (in1_uses_buffer) {
```
instead, create a unit test, and put up a PR for review.

Before that happens, @johanna-rock-tt @alingTT you can try that on your branch or use https://github.com/tenstorrent/tt-metal/pull/34662 (still in draft since I haven't tested this extensively and not added a unit test yet).
```
diff --git a/ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_1d_program_factory.cpp b/ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_1d_program_factory.cpp
index b61ce14d182..ec501181af3 100644
--- a/ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_1d_program_factory.cpp
+++ b/ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_1d_program_factory.cpp
@@ -2535,8 +2535,8 @@ inline void override_gather_in0_program_parameters(
             UpdateDynamicCircularBufferAddress(program, cb_output, *out_buffer);
         }
     }
-
-    if (not src1_sharded) {
+    bool in1_uses_buffer = !src1_sharded || (src1_sharded && src_buffer_b->is_dram());
+    if (in1_uses_buffer) {
         auto& writer_runtime_args_by_core = GetRuntimeArgs(program, override_variables.kernels.at(0));
         for (uint32_t i = 0; i < override_variables.cores.size(); ++i) {
             const auto& core = override_variables.cores[i];
```

**Comment by ekuslysTT** [comment]
also see https://github.com/tenstorrent/tt-metal/issues/34008#issuecomment-3670400995

**Comment by bbradelTT** [comment]
A workaround for the prefetch and a fix for the program cache has been merged in.

Should this issue be closed since accuracy is being looked at in #34008 ?

If not, what else needs to be done?

**Comment by alingTT** [comment]
I tested the fix and can confirm it resolves program caching issues for ring matmul, when I run accuracy tests with no prefetcher I get bad outputs with the fix provided all good outputs

**Comment by bbradelTT** [comment]
Can this issue be closed?

**Comment by alingTT** [comment]
Closing for now, if there are other issues with ring matmul we can file another issue
